src/router/builder/mod.rs

```diff
@@ -0,0 +1,448 @@
+//! Defines a builder API for constructing a `Router`.
+
+mod draw;
+mod single;
+mod replace;
+
+use std::marker::PhantomData;
+
+use hyper::StatusCode;
+
+use router::Router;
+use router::tree::TreeBuilder;
+use router::response::extender::ResponseExtender;
+use router::response::finalizer::ResponseFinalizerBuilder;
+use router::route::Delegation;
+use router::route::matcher::RouteMatcher;
+use router::route::dispatch::{PipelineHandleChain, PipelineSet};
+use router::request::path::PathExtractor;
+use router::request::query_string::QueryStringExtractor;
+use router::tree::node::NodeBuilder;
+
+pub use self::single::DefineSingleRoute;
+pub use self::draw::{DrawRoutes, DefaultSingleRouteBuilder};
+
+/// Builds a `Router` using the provided closure. Routes are defined using the `RouterBuilder`
+/// value passed to the closure, and the `Router` is constructed before returning.
+///
+/// ```rust
+/// # extern crate gotham;
+/// # extern crate hyper;
+/// # use hyper::Response;
+/// # use gotham::state::State;
+/// # use gotham::router::Router;
+/// # use gotham::router::builder::*;
+/// # use gotham::middleware::pipeline::new_pipeline;
+/// # use gotham::middleware::session::NewSessionMiddleware;
+/// # use gotham::router::route::dispatch::{new_pipeline_set, finalize_pipeline_set};
+/// # fn my_handler(_: State) -> (State, Response) {
+/// #   unreachable!()
+/// # }
+/// #
+/// fn router() -> Router {
+///     let pipelines = new_pipeline_set();
+///     let (pipelines, default) =
+///         pipelines.add(new_pipeline().add(NewSessionMiddleware::default()).build());
+///
+///     let pipelines = finalize_pipeline_set(pipelines);
+///
+///     let default_pipeline_chain = (default, ());
+///
+///     build_router(default_pipeline_chain, pipelines, |route| {
+///         route.get("/request/path").to(my_handler);
+///     })
+/// }
+/// # fn main() { router(); }
+/// ```
+pub fn build_router<C, P, F>(pipeline_chain: C, pipelines: PipelineSet<P>, f: F) -> Router
```

Comment:
I appreciate this PR has not considered pipelines so I'm happy for this comment to be considered in the future if overkill now.

I was considering use cases for Gotham and it strikes me that `pipelines` could very easily be something that developers who just want something simple could do without. Likewise in most of our API example documentation, where it becomes ~9 lines of boilerplate that could be left out.

How would you feel about adding a function like: 

`pub fn build_simple_router<F>(f: F) -> Router`

That internally does the basic pipeline setup before delegating to `build_router`? 

Moving between the two, if developers have some future use case for pipelines come up, would be a simple exercise as well.

Reply:
Sounds fine to me. I assume by "basic pipeline setup" you mean with no middleware?

Reply:
Indeed, as bare bones as it can be, essentially "just give me routes to my functions".

Reply:
For posterity: Sorted in 685e0d5